pkg/pillar/cmd/baseosmgr/handlebaseos.go

```diff
@@ -1018,3 +1023,169 @@ func handleOtherPartRebootReason(ctxPtr *baseOsMgrContext, status *types.BaseOsS
+func handleUpdateRetryCounter(ctxPtr *baseOsMgrContext, retryUpdateCounter uint32) {
+	curPartName := zboot.GetCurrentPartition()
+	partStatus := getZbootStatus(ctxPtr, curPartName)
+	if partStatus == nil {
+		log.Warnf("handleUpdateRetryCounter: No current partition status for %s; failed RetryUpdateCounter",
+			curPartName)
+		return
+	}
+	if partStatus.PartitionState != "active" {
+		if ctxPtr.configUpdateRetry == retryUpdateCounter {
+			log.Functionf("No change in retryUpdateCounter: %d", retryUpdateCounter)
+			return
+		}
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry changed "+
+			"with no active partition: %d to %d; ignoring it",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		return
+	}
+	if failed, failedPartStatus := isImageInErrorState(ctxPtr); failed {
+		if ctxPtr.configUpdateRetry == retryUpdateCounter {
+			log.Functionf("No change in retryUpdateCounter: %d", retryUpdateCounter)
+			return
+		}
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry change: %d to %d",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		ctxPtr.configUpdateRetry = retryUpdateCounter
+		saveCurrentRetryUpdateCounter(ctxPtr)
+		// save it to avoid loop of re-upgrade - failing - re-upgrade
+		saveConfigRetryUpdateCounter(ctxPtr)
+		log.Noticef("UpdateRetry from %s to %s",
+			partStatus.ShortVersion, failedPartStatus.ShortVersion)
+		zboot.SetOtherPartitionStateUpdating(log)
+		updateAndPublishZbootStatus(ctxPtr, failedPartStatus.PartitionLabel, false)
+		baseOsStatus := lookupBaseOsStatusByPartLabel(ctxPtr, failedPartStatus.PartitionLabel)
+		if baseOsStatus != nil {
+			baseOsSetPartitionInfoInStatus(ctxPtr, baseOsStatus, failedPartStatus.PartitionLabel)
+			publishBaseOsStatus(ctxPtr, baseOsStatus)
+		}
+		return
+	}
+	if ctxPtr.configUpdateRetry != retryUpdateCounter {
+		log.Noticef("handleUpdateRetryCounter: configUpdateRetry change: %d to %d",
+			ctxPtr.configUpdateRetry, retryUpdateCounter)
+		ctxPtr.configUpdateRetry = retryUpdateCounter
+		saveConfigRetryUpdateCounter(ctxPtr)
+	}
+	if ctxPtr.currentUpdateRetry != retryUpdateCounter {
+		log.Noticef("handleUpdateRetryCounter: currentUpdateRetry change: %d to %d",
+			ctxPtr.currentUpdateRetry, retryUpdateCounter)
+		ctxPtr.currentUpdateRetry = retryUpdateCounter
+		saveCurrentRetryUpdateCounter(ctxPtr)
+	}
+	publishBaseOSMgrStatus(ctxPtr)
+}
+
+func lookupBaseOsConfigByVersion(ctxPtr *baseOsMgrContext, shortVersion string) *types.BaseOsConfig {
+	sub := ctxPtr.subBaseOsConfig
+	items := sub.GetAll()
+	for _, cfg := range items {
+		baseOSConfig := cfg.(types.BaseOsConfig)
+		if baseOSConfig.BaseOsVersion == shortVersion {
+			return &baseOSConfig
+		}
+	}
+	return nil
+}
+
+func saveCurrentRetryUpdateCounter(ctxPtr *baseOsMgrContext) {
+	log.Functionf("saveCurrentRetryUpdateCounter (%s) - counter: %d",
+		currentRetryUpdateCounterFile, ctxPtr.currentUpdateRetry)
+	err := ioutil.WriteFile(currentRetryUpdateCounterFile,
```

Comment:
Shouldn't this use fileutils.WriteRename() to avoid issues with zero-lengh or partially written files if we have a power failure?

Reply:
Sure, done